cmd/zoekt-sourcegraph-indexserver/queue.go

```diff
@@ -36,6 +38,12 @@ type queueItem struct {
 	// dateAddedToQueue is the time when this indexing job was added to the queue. If this item is no longer
 	// in the heap (i.e. it has been processed already), this value is nonsensical.
 	dateAddedToQueue time.Time
+	// consecutiveFailures is the count of preceding consecutive failures. Backoff duration upon indexing failure increases linearly
+	// with this count,
+	consecutiveFailures int
+	// backOffUntil is the earliest time when we allow the item to be pushed to the heap. Until then the item will not be enqueued
+	// and indexing will not be attempted.
+	backoffUntil time.Time
```

Comment:
Another decision I made was to not enqueue if a repo is currently in backoff. We could enqueue and then at the time `processQueue` calls `Pop(`) we decide in the queue whether or not to return options or just discard and return error so that we don't process an item.

This might remove the additional logic on enqueue time.  However we'd be pushing items that might traverse the full height of the heap, only for it to be discarded on pop. An item that failed in its previous attempt(s) will be prioritized due to `indexed=false` being greater priority (see `Less()`). This seems wasteful.

Reply:
I think this is the correct decision, as long as we keep on updating the item's state we keep track of (index). Then when the item finally comes out of backoff bump will put it back into the queue.